### ISSUE TITLE
Evaluate "Calculate changed tiles by affected Shapes" Approach

### ISSUE THREAD

**Issue by lukasmartinelli** [osm2vectortiles/osm2vectortiles#129]
Manuel and me thought some more about the problem of detecting dirty tiles #16.
What we will try now is querying all changed shapes and calculating the affected tiles using https://github.com/mapbox/tile-cover. These affected tiles can then be rerendered.

We think this will be much faster than doing a DFS #123 and still perform if a million of geometries have changed. 

Below I listed two examples to clarify the approach.

### Detect Changes

In the ` osm_poi_points` new shapes have been inserted by imposm #122.
We can find out the changes by querying all shapes with a newer timestamp.

```sql
SELECT ST_AsGeoJSON(geometry) AS geometry
FROM osm_poi_points
WHERE timestamp > 'last-imported-timestamp'::timestamp
```

We now know that some `osm_poi_points` have been updated and we need to rerender parts of the world. We now need to find out which shapes have changed for which zoom levels (which depends on the queries in the tm2source project).

For `osm_poi_points` this is easy because POIs only appear on a single zoom level 14. We need to rerender tiles on zoom level 14 for the changed records.

### Calculate Tiles

For each changed geometry (in this case it is point and therefore fairly easy to calculate the affected tiles) we can now use tilecover to calculate the affected
tiles for zoom level 14.

```javascript
var geom = JSON.parse(geometry.geometry);
// Will only return a single tile for a point
var dirtyTiles = tileCover.getCoveredTiles(geom, 14);
```

Now we have an array of dirty tiles we need to rerender.

## How this works for changed Road Linestrings

Roads are a bit trickier because they are linestrings and behave differently on different zoom levels.

### Detect Changes

Let's find out all changes for zoom level 14, 13, and 12 (there are more affected ones). By querying streets we actually also solve the problem for tunnels and bridges.

**Affected geometries for z14**

```sql
SELECT ST_AsGeoJSON(geometry) AS geometry
FROM osm_roads
WHERE timestamp > 'last-imported-timestamp'::timestamp
```

**Affected geometries for z13**

```sql
SELECT ST_AsGeoJSON(geometry) AS geometry
FROM osm_roads
WHERE timestamp > 'last-imported-timestamp'::timestamp
 AND classify_road(type) IN ('minor_rail', 'street_limited', 'service')
```

**Affected geometries for z12**

```sql
SELECT ST_AsGeoJSON(geometry) AS geometry
FROM osm_roads
WHERE timestamp > 'last-imported-timestamp'::timestamp
 AND  classify_road(type) IN ('major_rail', 'street')
```

Now we a list of affected geometries for each zoom level.

### Calculate Tiles

For each changed geometry in zoom level 14.

```javascript
var dirtyTiles = tileCover.getCoveredTiles(geom, 14);
```
For each changed geometry in zoom level 13.

```javascript
var dirtyTiles = tileCover.getCoveredTiles(geom, 13);
```

For each changed geometry in zoom level 12.

```javascript
var dirtyTiles = tileCover.getCoveredTiles(geom, 12);
```

The beauty is that tile-cover will do the heavy lifting for us.

![Mapbox Tile Cover for Linestrings](https://camo.githubusercontent.com/457c6b6ae74500bf5a81407feb9c2b5b4fa7849b/68747470733a2f2f646c2e64726f70626f782e636f6d2f732f3438636a3136667674386e7968336f2f53637265656e73686f74253230323031342d30382d303625323031332e33342e31322e706e67)

**Comment by lukasmartinelli** [comment]
Approach works for more complex layers like landuse 4db5524ea9278c3ecbd738b2bfaab78990bf5ede. Perhaps we should implement some DSL to make it more readable.

**Comment by klokan** [comment]
We have talked about this approach on the roadmap meeting as well - the "give me changed shape and I tell you what tiles are dirty for it" way. In ideal case also on which zoom levels.
 
It seems you can generate the list of affected tiles directly in SQL as well - as I have mentioned already in comment in #123. See: http://javisantana.com/2014/10/22/traversing-quadtree.html This could run in the SQL trigger on import / update / delete of a feature. The example select is only for points - but the same principle can be applied on a single shape (polygon/line/point). You get the list of tiles on all zoom levels out the SQL command - and can save the list of tiles into a separate table.
For a single shape it could be very fast.

Would not this be better then calling for the same task an external node-based utility?

**Comment by lukasmartinelli** [comment]
> Would not this be better then calling for the same task an external node-based utility?

If it works as well as tile-cover it is much more elegant indeed (SQL :heart: ).
Will play with those queries as well and try to implement the same examples as proof of concept.
If we can adapt the query to work for polygons and lines in a reasonable amount of time doing everything on the database is much more elegant (performance wise I guess it will be about the same).

I'm a bit afraid it could end up in a big ball of SQL code but so far we actually managed to keep the SQL reasonably clean with the `pgsql` functions in the tm2source project so far.

Running it on the trigger or on demand  doesn't matter imho. But I would prefer explicitly querying for dirty tiles (and then a external tool can decide how to schedule it as jobs or print it as list etc) instead of keeping a table of dirty tiles in the database.

**Comment by klokan** [comment]
If the concept proves to be correct - the code could be rewritten to golang for direct integration in imposm3, before it enters PostGIS- with the list of dirty tiles directly stored in LevelDB cache. But let's make first the proof of concept in Node.js. Everything else is performance optimisation.

**Comment by lukasmartinelli** [comment]
I like the SQL solution. It is reasonably fast and worked on first try as well.
We just need to organize the code well. Below is example for `osm_poi_points`.

```sql
WITH RECURSIVE dirty_tiles(x, y, z, e) AS (
  -- root node (0, 0, 0)
  SELECT 0, 0, 0, EXISTS(
    SELECT 1 FROM changed_poi_points WHERE geometry && CDB_XYZ_Extent(0, 0, 0)
  )
  UNION ALL
  -- coordinate for the children
  SELECT x*2 + xx, y*2 + yy, z+1, EXISTS(
    SELECT 1 FROM changed_poi_points
    WHERE geometry && CDB_XYZ_Extent(x*2 + xx, y*2 + yy, z+1)
    ) FROM dirty_tiles,
    -- iterate over 4 children
    (VALUES (0, 0), (0, 1), (1, 1), (1, 0)) as c(xx, yy) 
    -- only for tiles with geometry and up to zoom level 14
    WHERE e AND z < 14
), changed_poi_points AS (
  SELECT * FROM osm_poi_points WHERE timestamp > '2015-10-01 21:23:01'::timestamp AND name <> ''
)
SELECT z, x, y FROM dirty_tiles where e;
```

On the other hand having it in Node is quite comfortable because rendering the missing tiles will also take place in Node.
I suggest we continue prototyping in Node (all we do is stitch together SQL queries in node - which is quite comfortable and productive so far). Porting it to pure SQL afterwards is straightforward (we only need to replace the "figure out dirty tiles from geometry part"). Lets discuss this tomorrow @manuelroth.

I also think it is important that we have a complete walkthrough the entire process asap so we can **actually** verify that it works with a concrete example.

> the code could be rewritten to golang for direct integration in imposm3,

We would need to hardcode the rules into Golang. Because now we really look at the queries of the tm2source project and build queries for each zoom level. The solution is not as abstract in the DFS approach. It all depends on the performance we will experience later down the road or until we can make more educated guesses and then start optimizing.
I am actually quite confident at the moment that figuring out the dirty tiles with the second approach is fast enough.

What is another idea (perhaps unnecessarily complicated - but definitly fun) is abstracting the tm2source queries into another format (Mapzen uses a Jade templating language to generate the tm2source queries - and you can solve any CS problem by another layer of indirection ;) 

Perhaps we can generate from that DSL both the queries for the tm2source and the queries to select the dirty tiles.

**Comment by lukasmartinelli** [comment]
Another idea how to write the "what features are displayed on what zoom" level only once and not in the queries **and** the dirty tiles logic.

We can create views for each zoom level and the layer is a `UNION ALL` query across all zoom level views.

## Example

Query for landuse before.

```sql
        (
          SELECT osm_id, geometry, classify_landuse(type) as class, type
            FROM (
              SELECT osm_id, geometry, type FROM osm_landusages_gen0
              WHERE geometry && !bbox!
              AND type IN ('wood', 'nature_reserve', 'national_park', 'forest')
              AND (
                (
                  st_area(geometry) > 300000000
                  AND z(!scale_denominator!) = 5
                )
                OR
                (
                  st_area(geometry) > 100000000
                  AND z(!scale_denominator!) = 6
                )
                OR
                (
                  st_area(geometry) > 25000000
                  AND z(!scale_denominator!) = 7
                )
                OR
                (
                  st_area(geometry) > 5000000
                  AND z(!scale_denominator!) = 8
                )
                OR
                (
                  st_area(geometry) > 2000000
                  AND z(!scale_denominator!) = 9
                )
                OR
                (
                  st_area(geometry) > 500000
                  AND z(!scale_denominator!) = 10
                )
              )
              UNION ALL
              SELECT osm_id, geometry, type FROM osm_landusages_gen1
              WHERE geometry && !bbox!
              AND (
                  type IN ('wood', 'nature_reserve', 'national_park', 'forest')
                  AND st_area(geometry) > 100000
              )
              AND z(!scale_denominator!) = 11
              UNION ALL
              SELECT osm_id, geometry, type FROM osm_landusages
              WHERE geometry && !bbox!
              AND (
                (
                  type IN ('wood', 'nature_reserve', 'national_park', 'forest')
                  AND st_area(geometry) > 10000
                  AND z(!scale_denominator!) = 12
                )
                OR
                (
                  type NOT IN ('wetland', 'marsh', 'swamp', 'bog', 'mud', 'tidalflat')
                  AND z(!scale_denominator!) >= 13
                )
              )
            ) AS landusages
        ) AS data
```

Query for landuse after. All logic for the zoom level distinction would be put into views. And we need to hope the query optimizer likes `UNION ALL`.

```sql
        (
          SELECT osm_id, geometry, classify_landuse(type) as class, type
            FROM (
              SELECT * FROM landuse_z5 WHERE z(!scale_denominator!) = 5
              UNION ALL
              SELECT * FROM landuse_z5 WHERE z(!scale_denominator!) = 6
              UNION ALL
              SELECT * FROM landuse_z7 WHERE z(!scale_denominator!) = 7
              UNION ALL
              SELECT * FROM landuse_z8 WHERE z(!scale_denominator!) = 8
              UNION ALL
              SELECT * FROM landuse_z9 WHERE z(!scale_denominator!) = 9
              UNION ALL
              SELECT * FROM landuse_z10 WHERE z(!scale_denominator!) = 10
              UNION ALL
              SELECT * FROM landuse_z11 WHERE z(!scale_denominator!) = 11
              UNION ALL
              SELECT * FROM landuse_z12 WHERE z(!scale_denominator!) = 12
              UNION ALL
              SELECT * FROM landuse_z13-z22 WHERE z(!scale_denominator!) >= 13
            ) AS landusages
            WHERE geometry && !bbox!
        ) AS data
```

And for the dirty tiles we can then specifically query the view only `SELECT * FROM landuse_z12 WHERE timestamp > ...`

**Comment by lukasmartinelli** [comment]
We now refactored the tm2source project to pull the data together from **zoom level specific views**. This has the benefit that we don't need to duplicate logic in order to query the dirty tiles.

For detecting the dirty tiles we are no longer using tile cover since the SQL scripts works so well. We use a dynamic SQL function `detect_dirty_tiles` that you can pass a timestamp and view to get back the dirty tile indizes.

This is very convenient to use for generating the dirty tiles. Querying the dirty tiles for the pois is a one liner.

```sql
SELECT * FROM detect_dirty_tiles('poi_label_z14', 'timestamp-of-last-change'::timestamp)
WHERE z = 14
```

We use this function from the Node script to write the dirty tiles to the list.

```javascript
function recentDirtyViews() {
    return db
        .one('SELECT MAX(timestamp) FROM osm_timestamps')
        .then(function(result) { return result.max; })
        .then(function(timestamp) {
            return q.all([
                detectDirtyTiles('water_label_z10', timestamp, 10, 10),
                detectDirtyTiles('water_label_z11', timestamp, 11, 11),
                detectDirtyTiles('water_label_z12', timestamp, 12, 12),
                detectDirtyTiles('water_label_z13', timestamp, 13, 13),
                detectDirtyTiles('water_label_z14', timestamp, 14, 14),
                detectDirtyTiles('poi_label_z14', timestamp, 14, 14),
                detectDirtyTiles('road_label_z8toz10', timestamp, 8, 10),
                detectDirtyTiles('road_label_z11', timestamp, 11, 11),
                detectDirtyTiles('road_label_z12toz13', timestamp, 12, 13),
                detectDirtyTiles('road_label_z14', timestamp, 14, 14),
                detectDirtyTiles('waterway_label_z8toz12', timestamp, 8, 12),
                detectDirtyTiles('waterway_label_z13toz14', timestamp, 13, 14),
                detectDirtyTiles('housenum_label', timestamp, 14, 14)
            ]);
        }).then(_.flatten);
}
```

**Comment by lukasmartinelli** [comment]
@manuelroth We need to add the views to the `drop_views` function
in  the `import-osm` container. Otherwise imposm cannot remove the tables that are referenced by the views.

https://github.com/osm2vectortiles/osm2vectortiles/blob/feature/detect-dirty-tiles/src/import-osm/import.sh#L81-L83

**Comment by manuelroth** [comment]
@lukasmartinelli Thanks for the info. Added them with commit https://github.com/osm2vectortiles/osm2vectortiles/commit/9a2927d053273a10d0ff222b79666d3f61e89a7a

**Comment by lukasmartinelli** [comment]
Yes it is a bit nasty but i found no other way to do it because imposm3 drops the tables.
This also means that after each `docker-compose import-osm` you have to run `docker-compose run import-sql`. But perhaps we can find a better way.

**Comment by lukasmartinelli** [comment]
Resolved by using tilecover in #154
